### PR TITLE
New module: 10x Space Ranger

### PR DESCRIPTION
**Pull request by grst**
- [x] This comment contains a description of changes (with reason)
- [x] `CHANGELOG.md` has been updated

- [x] There is example tool output for tools in the <https://github.com/ewels/MultiQC_TestData> repository or attached to this PR
    - [x] pending https://github.com/ewels/MultiQC_TestData/pull/256
- [x] Code is tested and works locally (including with `--lint` flag)
- [x] `docs/README.md` is updated with link to below
- [x] `docs/modulename.md` is created
- [x] Everything that can be represented with a plot instead of a table is a plot
- [x] Report sections have a description and help text (with `self.add_section`)
- [ ] There aren't any huge tables with > 6 columns (explain reasoning if so)
   - similar to cellranger there are many relevant metrics that are best represented as table
   - I tried to hide less common ones by default. 
- [x] Each table column has a different colour scale to its neighbour, which relates to the data (eg. if high numbers are bad, they're red)
- [x] Module does not do any significant computational work

---

Add a module for [10x Space Ranger](https://support.10xgenomics.com/spatial-gene-expression/software/pipelines/latest/what-is-space-ranger). 

Since the output reports are very similar (yet not identical) to the Cellranger reports, I started by copying the cellranger module.

### CODE REVIEW

**Comment by ewels** [comment]
For next time (I know you copied this from the CellRanger module so this isn't really anything to do with you @grst, but still I feel compelled to say) - please don't make MultiQC module code too clever 😅 

I'm trying to tweak the General Stats header config and I'm spending ages going around in circles trying to find where the attributes are set - in which function, which file, which dict. It's really much better to have verbose repetitive code that is easy to read, standardised with other MultiQC modules, and trivial to customise on a per-column basis. Even though everyone's knee-jerk response to seeing repetitive code is to optimise into a function, everyone does it differently and it hinders customisation - it causes a lot of hassle.

Rant / shouting into the void over... 😬

**Comment by ewels** [comment]
Mostly looking good, going from the report - thanks for this!

Quick thoughts:
* Can we chop the decimal off a bunch of the table values that are integers (gene counts, spot counts etc)
* In the table with pass / fail status - what are the blanks? Are they passes? Can they be marked as such?
* If we could check that the line charts data being downsampled for the plot, so that each line doesn't have 11k data points in the report and creates a massive filesize. See config options `smooth_points` ([docs](https://multiqc.info/docs/development/plots/#line-graphs))

**Comment by grst** [comment]
Hi @ewels, 

I agree with pretty much everything of your rant. I needed a lot of time myself to understand the logic of the Spaceranger module and was particularly unhappy with the way Mixture classes are used here -- I still figured it was the fastest way to get the job done to reuse the code. 

Regarding your specific feedback: 
> Can we chop the decimal off a bunch of the table values that are integers (gene counts, spot counts etc)

should be fixed

> In the table with pass / fail status - what are the blanks? Are they passes? Can they be marked as such?

the problem is that the cellranger/spaceranger HTML reports do not report passes, only failures. So yes, we could mark them as passes, but we never get the full matrix, only the samples that have at least one failure and only the failures that appear in at least one sample show up in the table. I'd be included to leave it as-is, but happy to change it if you prefer. 

> If we could check that the line charts data being downsampled for the plot

I checked the JSON we extract from the spaceranger report and don't think any further downsampling is necessary. It's about 20 data points per sample.

**Comment by grst** [comment]
@ewels @vladsavelyev, I fixed the tests again.

**Review by vladsavelyev — APPROVED** [review]
Merged with main, some minor clean-up, and good to merge!

File: multiqc/modules/spaceranger/count.py [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+        # Extract warnings if any
+        warnings = {}
+        alarms_list = summary["alarms"].get("alarms", [])
+        for alarm in alarms_list:
+            # "Intron mode used" alarm added in Space Ranger 7.0 lacks id
+            if "id" not in alarm:
+                continue
+            warnings[alarm["id"]] = "FAIL"
+            self.count_warnings_headers[alarm["id"]] = {
+                "title": alarm["id"],
```
[the remark this answers is not in the document]
Comment (vladsavelyev) [reply multiqc/modules/spaceranger/count.py:267]: Done

**Pull request merged**